Unfortunate: the the un4chan userscript toolkit

qqueue opened the pull request:
i.e., the non-ui parts of c4, so others can userscript without being tied to either 4chan's UI choices or c4's UI choices (as much as possible, anyway). `unf` for short.

I aiming for this library to be usable with [polymer](https://www.polymer-project.org/), so one could make a Material Design(tm) userscript on top. So far so good, though polymer itself might take some convincing to work within greasemonkey though (haven't tried it yet).

Still TODO is extracting the fancy new SharedWorker-based updater into something usable. I basically want to have an API like:

```
unf.data.subscribe board, tno, (thread) ->
  ui.rerender-with-new-data thread
```

which should plug in to most MVWhatever libraries (polymer, react, mithril, etc) nicely. Then, just wrap this up in a separate buildscript/repository, add npm + publishing for greasemonkey `@require`, and it's ready for use, albeit probably buggy (like c4).

There are some other parts of c4 that might be worth extracting as well, such as the relative dates, lightbox/tooltips, the youtube api queue, and the posting logic. Once that's extracted, c4 is a pretty tiny codebase: some html templates, styles, and a bundle of hideous dom-mangling.

@saxamaphone69 , 4u.

qqueue committed:
Unfortunate: port first API

happening.jpg

qqueue committed:
Unfortunate: port onready

lots more ugliness both within c4 and
the interface I just invented. It'll get better soon though

vendethiel commented on src/unfortunate/parser.co:
```diff
@@ -0,0 +1,177 @@
+
+  op = new DOMPost do
+    thread-no
+    el.querySelector( \.op )
```
why the spacing? 10 lines below you don't space it like that anymore

vendethiel replied:
™

vendethiel commented on src/unfortunate/index.co:
```diff
@@ -0,0 +1,116 @@
+export function locate location
+  # boards.4chan.org/<name>/'thread' or <page>/<threadid>, or 'catalog'/<searchterm>
+  [ , board, type, qualifier] = location.pathname / \/
+  if type is \catalog
```
switch?

vendethiel commented on src/unfortunate/index.co:
```diff
@@ -0,0 +1,116 @@
+export function parse-page document
+  if document.querySelector \#threads # it's a catalog
+    # find script tag that defines `catalog`
+    [catalog-text] = Array::filter.call document.querySelectorAll(\script), ->
```
`[...$$(\script)].filter ->` works as well (if you're in for moving your `$` and `$$` and `L` helpers to some other libraries)

vendethiel commented:
as I said - why not move your helpers-thingies into some npm module as well, and require it both from unfortunate and from the other one?

saxamaphone69 commented:
mfw i still have no idea what this means, but am excited nonetheless

vendethiel commented:
You can expect great things in @qqueue... You just need to wait a bit : d

qqueue committed:
Unfortunate: port updater

Couldn't think of a nice api, so I just wrapped the current
api up in new naming conventions. Good enough for testing, I think.

qqueue commented on src/unfortunate/parser.co:
```diff
@@ -0,0 +1,177 @@
+
+  op = new DOMPost do
+    thread-no
+    el.querySelector( \.op )
```
legacy whitespace. I'm pretty sure the DOM parser is the oldest still-active piece of code in c4/html5chan.

I'll clean it up soon(tm)

qqueue commented:
> why not move your helpers-thingies into some npm module as well

I might, but after looking through the code several times, the helpers really aren't that much helpful, and they make the source ever-so-slightly-harder to read without familiarizing yourself with my idiosyncratic set of symbol-named functions. I _might_ extract `timing.co`, but the other helper/utils seem pretty useless outside of the "hideous dom mangling" done in `inlinereplies.co`.

qqueue commented:
After much wrangling, I managed to get a `<paper-button>` from Polymer working within greasemonkey. I used [vulcanize][0] to get rid of the imports, then manually concatted the rest of the javascript into this:

https://gist.github.com/qqueue/0489a75a10caba0746f4

The userscript block (my code) starts at 18397 . Since it's running at `document-start` like c4 does, I just replaced the existing document.body and style with the output from vulcanize.

It didn't _quite_ work at first, but after liberal application of `console.log`, I figured out that the weirdness around Greasemonkey's `window` and `this` context confuse some parts of the polymer library code that somehow define `window.Polymer` as function and `var Polymer`/`this.Polymer` as an object, presumably for legacy reasons. Changing `window.Polymer` to `this.Polymer` made everything (so far) work fine.

This means that my grand plan for polymer+unfortunate is likely feasible. I'll still need to to dig in to vulcanize and figure out how to output it in a userscript-friendly (styles as text, scripts inline) format, but fixing it to work with greasemonkey was at least a little easier than expected.

saxamaphone69 commented:
Interesting you got it working, the Polymer website doesn't even load correctly for me.

![browser-2014-12-18](https://cloud.githubusercontent.com/assets/877480/5479660/7bc0212a-8693-11e4-83d8-ee4b473a62e0.png)

And that's on Nightly. Maybe it's finally time to start using Chrome (isn't it easier to create "apps" too using it)?

vendethiel commented:
Fuck chrome.

saxamaphone69 commented:
oh ok sorry :disappointed:

qqueue commented:
I'm getting the same breakage on polymer's site, though only if I have `dom.webcomponents.enabled` set to `true` in about:config. Do you have that turned on @saxamaphone69 ?

Re: chrome extensions, I've always planned to make c4 work in chrome eventally (I did have it working in tampermonkey at one point), but since my userbase has always been just me, I don't have a lot of incentive. One thing i do have incentive for is moving out of greasemonkey and into the jetpack addon API (i.e., firefox extension), which seems to have a lot better debug tooling than greasemonkey. I could also then intercept 4chan page loads in a less hacky manner, and make my SharedWorker use an actual file instead of a data URI. https://github.com/mozilla/jpm is probably worth a try.

One other spec I've been keeping track of lately is ServiceWorkers, which allows you to intercept requests for a given domain and mangle the responses however you want, all from a separate javascript worker thread. If you think about it, that's pretty much what I've turned greasemonkey into with the DOM replacement trick--given a URL and the original response, output whatever you want with the old content as input. While it's not built for site augmentation like greasemonkey/addon api is, it does have some interesting parallels.

saxamaphone69 commented:
Flipped the pref and the site now works. I must have set it when it first got announced for Firefox and never bothered reading up on it or something.

I only mention Chrome because as far as I know, you can actually debug and test a lot of things within Chrome's console and stuff. But, I'm not an expert. A solo Firefox add-on might be cool, as I would assume doing so would also allow you greater control and such. The only reason I mention Chrome is because from memory, you don't need to worry about fancy stuff, and you can pretty much create one with straight HTML/CSS/JS? I don't know. You guys are the expert, I just like sitting and watching, as I've said before. :hamburger:

vendethiel commented:
@qqueue Please note that there's a 5$ fee to upload extensions to the store, and that "vanilla chrome" can't install non-store extensions (even in dev. mode). You need chromium or chrome canary to dev extensions. Yes, this is awful.

qqueue committed:
Parser: fix subject truncation

annoying, they must've changed the way that works.

vendethiel commented:
Just noticed the PR title has "the" twice in it.

qqueue committed:
catalog: fix to new catalog.json

breaking my ancient crufty shit. The ordering
no longer works, oh well.

qqueue committed:
catalog: specify thumbnail size

if you just wait 3 years, TODOs just get implemented, yo

vendethiel commented:
:+1:

qqueue committed:
Support /tg/ threads with pdfs

Since I now browse that sad board.

Holy shit is livescript sugary, I can see why I liked it.

qqueue committed:
fix google reverse img search url

qqueue committed:
Show img byte size and dimensions by default

People are always posting fuck huge images in /tg/ for
some reason, like 6 MB phone pics. annoying.

qqueue committed:
updater: fix dumb ff57 shit

They finally did it. They broke my 2-year-old shit script.
xhr no longer takes relative "//" urls anymore i guess,
so https everywhere. I'm kind of surprised my sharedworker
shit still works. I remember how terrible it all seemed when
I wrote it. It probably still sux. javascript life.

qqueue committed:
fix thumbnails moving to i.4cdn.org

hiro pls

vendethiel commented:
i thought something was actually happening x(

qqueue committed:
fix catalog-thread hardcoded t.4cdn

sad.
